# 01 - Data Overview

This notebook explores the raw Traffy dataset (bangkok_traffy.csv):
- check basic structure (rows, columns, dtypes)
- inspect important columns (type, district, state, timestamp, etc.)
- check missing values
- create a 100-row sample for sharing (`data_samples/00-sample_bangkok_traffy.csv`)

In [15]:
import pandas as pd

In [16]:
df = pd.read_csv("../data/bangkok_traffy.csv")

In [17]:
df.sample(100, random_state=42).to_csv("../data_samples/00-sample_bangkok_traffy.csv", index=False)

In [18]:
df_sample = pd.read_csv("../data_samples/00-sample_bangkok_traffy.csv")

In [19]:
df_sample.tail(1)

ticket_id                  type  \
99  2024-KHBWP7  {แสงสว่าง,ร้องเรียน}   

                                         organization  \
99  เขตพญาไท,สำนักการโยธา กทม.,ฝ่ายโยธา เขตพญาไท,ส...   

                                              comment  \
99  เสาไฟฟ้าหน้าบ้านดับ มากกว่า2เดือนแล้ว และบ้านข...   

                                                photo photo_after  \
99  https://storage.googleapis.com/traffy_public_b...         NaN   

                coords                                            address  \
99  100.53858,13.78434  208 ถ. พระรามที่ 6 แขวงพญาไท เขตพญาไท กรุงเทพม...   

   subdistrict district       province                      timestamp  \
99    สามเสนใน    พญาไท  กรุงเทพมหานคร  2024-12-26 13:17:20.258055+00   

             state  star  count_reopen                  last_activity  
99  กำลังดำเนินการ   NaN             0  2025-01-02 04:39:05.644133+00

OVERVIEW RAW DATA

In [ ]:
print("Shape:", df.shape)

Shape: (787026, 16)


In [33]:
df.sample(5,random_state=87)

ticket_id                   type  \
484101  2024-4CGK9N                  {ถนน}   
259468       CGJVPW                  {ถนน}   
514015  2024-G63ER3      {ความปลอดภัย,ถนน}   
583260  2024-ELG3R9          {ท่อระบายน้ำ}   
643813  2024-GBGU2C  {คลอง,เสียงรบกวน,ถนน}   

                                             organization  \
484101  เขตสะพานสูง,ฝ่ายเทศกิจ เขตสะพานสูง,ฝ่ายโยธา เข...   
259468                 เขตคลองสามวา,ฝ่ายโยธา เขตคลองสามวา   
514015                         เขตบางนา,ฝ่ายโยธา เขตบางนา   
583260                         เขตวัฒนา,ฝ่ายโยธา เขตวัฒนา   
643813  เขตคันนายาว,สน.บางชัน,ฝ่ายเทศกิจ เขตคันนายาว,ฝ...   

                                                  comment  \
484101  มีเศษหิน ดิน ทราย หล่นบนพื้นถนน ออกมาจากพื้นที...   
259468  ถนนหทัยราษฎร์ เขตคลองสามวา \nบนถนนดังกล่าว (จุ...   
514015  ถนน เป็นหลุมเป็นบ่อ อาจเกิดอันตรายแก่ผู้ที่สัญ...   
583260                                     ท่อระบายน้ำตัน   
643813  ตรงซอยเสรีไทย52 แยกไปทางขวา ถนนเลียบคลองหลอแหล...   

                                                    photo  \
484101  https://storage.googleapis.com/traffy_public_b...   
259468  https://storage.googleapis.com/traffy_public_b...   
514015  https://storage.googleapis.com/traffy_public_b...   
583260  https://storage.googleapis.com/traffy_public_b...   
643813  https://storage.googleapis.com/traffy_public_b...   

                                              photo_after              coords  \
484101  https://storage.googleapis.com/traffy_public_b...  100.69395,13.74095   
259468  https://storage.googleapis.com/traffy_public_b...  100.72258,13.86010   
514015  https://storage.googleapis.com/traffy_public_b...  100.59781,13.67497   
583260                                                NaN  100.59686,13.71230   
643813  https://storage.googleapis.com/traffy_public_b...  100.69743,13.79477   

                                                  address   subdistrict  \
484101  8 ซอย นักกีฬาแหลมทอง แขวงสะพานสูง เขตสะพานสูง ...      สะพานสูง   
259468  ถนนหทัยราษฎร์ คลองสามวา กรุงเทพมหานคร : VP6F+6...        บางชัน   
514015  98/16 ซอย กุศลศิลป์ แขวง บางนา เขตบางนา กรุงเท...      บางนาใต้   
583260  989 ถ. สุขุมวิท แขวงพระโขนงเหนือ เขตวัฒนา กรุง...  พระโขนงเหนือ   
643813  8/150 มบ.กลางเมือง ซ. เสรีไทย 52 แขวงคันนายาว ...      คันนายาว   

         district              province                      timestamp  \
484101   สะพานสูง         กรุงเทพมหานคร  2024-01-23 05:20:45.707202+00   
259468  คลองสามวา  จังหวัดกรุงเทพมหานคร  2023-04-25 08:02:51.466057+00   
514015      บางนา         กรุงเทพมหานคร  2024-02-29 02:58:30.640753+00   
583260      วัฒนา         กรุงเทพมหานคร  2024-05-29 08:42:59.622794+00   
643813   คันนายาว         กรุงเทพมหานคร  2024-08-03 13:03:17.897463+00   

                 state  star  count_reopen                  last_activity  
484101       เสร็จสิ้น   1.0             3  2024-02-15 03:19:27.440436+00  
259468       เสร็จสิ้น   NaN             0  2023-04-27 04:40:59.570554+00  
514015       เสร็จสิ้น   5.0             0  2024-06-12 04:40:01.768992+00  
583260  กำลังดำเนินการ   NaN             0  2024-05-29 11:09:31.859006+00  
643813  กำลังดำเนินการ   NaN             0  2024-08-14 06:53:31.365103+00

In [35]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 787026 entries, 0 to 787025
Data columns (total 16 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   ticket_id      778254 non-null  object 
 1   type           786929 non-null  object 
 2   organization   786455 non-null  object 
 3   comment        778254 non-null  object 
 4   photo          786911 non-null  object 
 5   photo_after    641309 non-null  object 
 6   coords         787026 non-null  object 
 7   address        778254 non-null  object 
 8   subdistrict    786460 non-null  object 
 9   district       786465 non-null  object 
 10  province       786831 non-null  object 
 11  timestamp      787026 non-null  object 
 12  state          787026 non-null  object 
 13  star           274097 non-null  float64
 14  count_reopen   787026 non-null  int64  
 15  last_activity  787026 non-null  object 
dtypes: float64(1), int64(1), object(14)
memory usage: 96.1+ MB


In [38]:
df.describe()

star   count_reopen
count  274097.000000  787026.000000
mean        3.872801       0.123485
std         1.471863       0.823554
min         1.000000       0.000000
25%         3.000000       0.000000
50%         5.000000       0.000000
75%         5.000000       0.000000
max         5.000000     100.000000

In [44]:
missing = df.isna().sum().sort_values(ascending=False)
missing = missing.to_frame(name="missing_count")
missing["missing_ratio"] = missing["missing_count"] / len(df)
missing

missing_count  missing_ratio
star                  512929       0.651731
photo_after           145717       0.185149
ticket_id               8772       0.011146
comment                 8772       0.011146
address                 8772       0.011146
organization             571       0.000726
subdistrict              566       0.000719
district                 561       0.000713
province                 195       0.000248
photo                    115       0.000146
type                      97       0.000123
coords                     0       0.000000
timestamp                  0       0.000000
state                      0       0.000000
count_reopen               0       0.000000
last_activity              0       0.000000

In [53]:
print("unique")
print("type",df["type"].nunique())
print("province",df["province"].nunique()) 
print("district",df["district"].nunique())
print("subdistrict",df["subdistrict"].nunique())
print("organization",df["organization"].nunique())
print("state",df["state"].nunique(),df["state"].unique())


unique
type 17469
province 81
district 144
subdistrict 372
organization 70208
state 3 ['เสร็จสิ้น' 'กำลังดำเนินการ' 'รอรับเรื่อง']


In [55]:
df["timestamp"] = pd.to_datetime(df["timestamp"], format="mixed", errors="coerce")

print("min",df["timestamp"].min())
print("max",df["timestamp"].max())

min 2021-09-03 12:51:09.453003+00:00
max 2025-01-16 02:53:34.290375+00:00


In [56]:
df["year"] = df["timestamp"].dt.year
df["year"].value_counts().sort_index(ascending=False)

year
2025     12719
2024    307931
2023    276399
2022    189965
2021        12
Name: count, dtype: int64

In [59]:
df["comment_length"] = df["comment"].fillna("").str.len()
df["comment_length"].describe()

count    787026.000000
mean        169.449712
std         254.983385
min           0.000000
25%          39.000000
50%          96.000000
75%         217.000000
max       10620.000000
Name: comment_length, dtype: float64

In [60]:
df.head(1)

ticket_id         type organization  comment  \
0  2021-FYJTFP  {ความสะอาด}   เขตบางซื่อ  ขยะเยอะ   

                                               photo photo_after  \
0  https://storage.googleapis.com/traffy_public_b...         NaN   

               coords                                            address  \
0  100.53084,13.81865  12/14 ถนน กรุงเทพ- นนทบุรี แขวง บางซื่อ เขตบาง...   

  subdistrict district       province                        timestamp  \
0         NaN      NaN  กรุงเทพมหานคร 2021-09-03 12:51:09.453003+00:00   

       state  star  count_reopen                  last_activity  year  \
0  เสร็จสิ้น   NaN             0  2022-06-04 15:34:14.609206+00  2021   

   comment_length  
0               7

## Summary — Raw Traffy Data Overview

### 1) Dataset Snapshot
- File loaded: `../data/bangkok_traffy.csv`
- Shape: **787,026 rows**, **16 columns**
- Columns:
  - `ticket_id`, `type`, `organization`, `comment`
  - `photo`, `photo_after`
  - `coords`, `address`, `subdistrict`, `district`, `province`
  - `timestamp`, `state`, `star`, `count_reopen`, `last_activity`

### 2) Column Types and Structure
- `df.info()` shows:
  - 14 columns are `object` (string-like)
  - 1 column is `float64` → `star`
  - 1 column is `int64` → `count_reopen`
- `timestamp` and `last_activity` are initially `object` and later converted to datetime
- Memory usage ~96 MB

### 3) Missing Values Overview
Top missing columns (from `missing` table):
- `star`: **512,929 missing** (~65.17%)
- `photo_after`: **145,717 missing** (~18.51%)
- `ticket_id`, `comment`, `address`: **8,772 missing each** (~1.11%)
- `organization`: 571 missing (~0.07%)
- `subdistrict`: 566 missing (~0.07%)
- `district`: 561 missing (~0.07%)
- `province`: 195 missing (~0.02%)
- `photo`: 115 missing (~0.01%)
- `type`: 97 missing (~0.01%)
- No missing values in:
  - `coords`, `timestamp`, `state`, `count_reopen`, `last_activity`

### 4) Uniqueness and Categorical Complexity
From the unique counts:
- `type`: **17,469** unique values  
  → very high cardinality, many combinations like `{คลอง,เสียงรบกวน,ถนน}` → must standardize later
- `province`: 81 unique
- `district`: 144 unique
- `subdistrict`: 372 unique
- `organization`: **70,208** unique (very high, likely too granular for direct modeling)
- `state`: 3 unique values:
  - `'เสร็จสิ้น'`
  - `'กำลังดำเนินการ'`
  - `'รอรับเรื่อง'`

### 5) Time Range and Year Distribution
After converting `timestamp` with:

```python
df["timestamp"] = pd.to_datetime(df["timestamp"], format="mixed", errors="coerce")
